File: Data Structures & Algorithms/evaluate-reverse-polish-notation/submission-1.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for t in tokens:
            match t:
                case "+":
                    n1 = stack.pop() 
                    n2 = stack.pop() 
                    stack.append(n2 + n1)
                case "-":
                    n1 = stack.pop()
                    n2 = stack.pop()
                    stack.append(n2 - n1)
                case "*":
                    n1 = stack.pop()
                    n2 = stack.pop()
                    stack.append(n2 * n1)
                case "/":
                    n1 = stack.pop()
                    n2 = stack.pop()
                    stack.append(int(n2 / n1))
                case _:
                    stack.append(int(t))
                    
        return stack[0]

File: Data Structures & Algorithms/decode-string/submission-1.ts
class Solution {
    /**
     * @param {string} s
     * @return {string}
     */
    decodeString(s: string): string {

        const char = [];
        const num = [];

        let cur = '',
            k = 0; // multiplier 

        for (const ch of s) {
            if (ch >= '0' && ch <= '9') k = k * 10 + Number(ch);
            else if (ch === '[') {
                // push both to stack
                num.push(k);
                char.push(cur);
                // reset the current string and multiplier
                cur = '';
                k = 0;
            }
            else if (ch === ']') {
                const temp = cur;
                cur = char.pop()!;
                const count = num.pop()!;
                cur += temp.repeat(count);
            }
            else cur += ch;
        }
        return cur;
    }
}
